### PR TITLE
docs: add how-to-write-a-module.md

### PR DESCRIPTION
**Pull request by ilyam8**
https://i.pinimg.com/originals/fe/8f/26/fe8f26548e9c75575a0e0d25a88e58e4.jpg

### CODE REVIEW

**Comment by ilyam8** [comment]
@OdysLam ping

File: docs/how-to-write-a-module.md [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+ 
+-   module registration.
+-   module configuration.
+-   [module interface implementation](#module-interface).
+
+### File `charts.go`
+
+> :exclamation: See the example: [`charts.go`](https://github.com/netdata/go.d.plugin/blob/master/modules/example/charts.go).
+
+Put charts, charts templates and charts constructor functions in this file.
```
[the remark this answers is not in the document]
Comment (ilyam8) [reply docs/how-to-write-a-module.md:173]: > See the example: [`charts.go`]

[the remark this answers is not in the document]
Comment (OdysLam) [reply docs/how-to-write-a-module.md:173]: I think the document should be all-encompassing, meaning that I don't need to go here and there to get required information to complete my understanding. Furthermore, I still can't understand what those data structure are or what is their purpose. 
 Let's get as explicit as possible, ELI5 level.

[the remark this answers is not in the document]
Comment (OdysLam) [reply docs/how-to-write-a-module.md:173]: Our goal is to lower the barrier of entry as much as possible, so that people can contribute easier and with less required technical understanding.

[the remark this answers is not in the document]
Comment (ilyam8) [reply docs/how-to-write-a-module.md:173]: We expect people checking other modules source code and source code in general (it is very easy to do useing IDE these days), it is impossobile ot avoid it.

You think that this is difficult, but it is not.

[the remark this answers is not in the document]
Comment (OdysLam) [reply docs/how-to-write-a-module.md:178]: I checked code you mean, but what I can't understand is what those data structures means for module, from a logical perspective.

The idea is to allow even rookie go programmers to easily grasp the fundamentals and create modules, otherwise, people will continue using python.

[the remark this answers is not in the document]
Comment (OdysLam) [reply docs/how-to-write-a-module.md:178]: I checked the file. I want us to explain to users what those mean in the netdata context, what is a chart and what is a template. Why in the example, one is hidden, what's the functionality of a hidden chart (i.e why would I want to do it). 
Give me all the information so that I don't need to make **any** assumption. Allow me to focus on business logic and now to understand the netdata aspects.

File: docs/how-to-write-a-module.md [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+func (e *Example) collect() (map[string]int64, error) {
+    collected := make(map[string])int64
+    // ...
+    // ...
+    // ...
+    return collected, nil
+}
+```
+
+### File `module_name_test.go`
```
Comment (OdysLam) [inline docs/how-to-write-a-module.md:228]: Can we add some links to resources for creating tests in go?  Can we give some examples from already implemented collectors? 
Examples would be test code + explanation/description of what it does. I want even the go rookie programmer who hasn't written any tests to quickly understand the concept/logic and create the tests without giving up on the collector.
Remember, we want to nudge people from creating a quick and dirty python collector to using go which is more complex but more robust as well. 
Perhaps @kaskavel can help with this one?

File: docs/how-to-write-a-module.md [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+Testing is mandatory.
+
+-   test only public functions and methods (`New`, `Init`, `Check`, `Charts`, `Cleanup`, `Collect`).
+-   do not create a test function per a case, use [table driven tests](https://github.com/golang/go/wiki/TableDrivenTests).
+Prefer `map[string]struct{ ... }` over `[]struct{ ... }`.
+-   use helper functions _to prepare_ test cases to keep them clean and readable.  
+
+### Directory `testdata/`
+
+The directory contains tests fixtures.
```
Comment (OdysLam) [inline docs/how-to-write-a-module.md:241]: What are tests fixtures? Let's assume no knowledge from the part of the user.

File: docs/how-to-write-a-module.md (hunk 2 above)
Comment (kaskavel) [reply docs/how-to-write-a-module.md:228]: @OdysLam I will take a look now on this. Let me go through the existing documentation and I 'll post here.

**Comment by ilyam8** [comment]
@OdysLam in case you are using `Files changed` i suggest to look at the end result - https://github.com/ilyam8/go.d.plugin/blob/how-to-write-a-collector/docs/how-to-write-a-module.md.

File: docs/how-to-write-a-module.md [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+| [`module_name.go`](#file-module_namego)           | Module configuration, implementation and registration.  |
+| [`charts.go`](#file-chartsgo)                     | Charts, charts templates and constructor functions.     |
+| [`init.go`](#file-initgo)                         | Initialization methods.                                 |
+| [`collect.go`](#file-collectgo)                   | Metrics collection implementation.                      |
+| [`module_name_test.go`](#file-module_name_testgo) | Public methods/functions tests.                         |
+| [`testdata/`](#file-module_name_testgo)           | Files containing sample data.                           |
+
+### File `module_name.go`
+
+> :exclamation: See the example [`examlpe.go`](https://github.com/netdata/go.d.plugin/blob/master/modules/example/example.go).
```
Comment (OdysLam) [inline docs/how-to-write-a-module.md:164]:
```suggestion
> :exclamation: See the example [`example.go`](https://github.com/netdata/go.d.plugin/blob/master/modules/example/example.go).
```

File: docs/how-to-write-a-module.md [hunk 5]
```diff
@@ -0,0 +1,253 @@
+# How to write a module
+
+Let's assume you want to write `example` module.
```
Comment (joelhans) [inline docs/how-to-write-a-module.md:3]:
```suggestion
Let's assume you want to write a collector named `example`.
```

File: docs/how-to-write-a-module.md [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+Steps are:
+
+- add the source code to the [`modules/example/`](https://github.com/netdata/go.d.plugin/tree/master/modules).
+  - [module interface](#module-interface).
+  - [suggested module layout](#module-layout).
+  - [helper packages](#helper-packages).
+- add the configuration to the [`config/go.d/exmaple.conf`](https://github.com/netdata/go.d.plugin/tree/master/config/go.d).
+- add the module to the [`config/go.d.conf`](https://github.com/netdata/go.d.plugin/blob/master/config/go.d.conf).
+- import the module in [`modules/init.go`](https://github.com/netdata/go.d.plugin/blob/master/modules/init.go).
+- update the [`available modules list`](https://github.com/netdata/go.d.plugin#available-modules).
```
Comment (joelhans) [inline docs/how-to-write-a-module.md:14]:
```suggestion
- Add the source code to [`modules/example/`](https://github.com/netdata/go.d.plugin/tree/master/modules).
  - [module interface](#module-interface).
  - [suggested module layout](#module-layout).
  - [helper packages](#helper-packages).
- Add the configuration to [`config/go.d/example.conf`](https://github.com/netdata/go.d.plugin/tree/master/config/go.d).
- Add the module to [`config/go.d.conf`](https://github.com/netdata/go.d.plugin/blob/master/config/go.d.conf).
- Import the module in [`modules/init.go`](https://github.com/netdata/go.d.plugin/blob/master/modules/init.go).
- Update the [`available modules list`](https://github.com/netdata/go.d.plugin#available-modules).
```

File: docs/how-to-write-a-module.md [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+```go
+// example.go
+
+func (Example) Cleanup() {}
+```
+
+## Module Layout
+
+The general idea - do not put everything in a file.
```
Comment (joelhans) [inline docs/how-to-write-a-module.md:147]:
```suggestion
The general idea is to not put everything in a single file.
```

File: docs/how-to-write-a-module.md [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+
+We recommend using a file per a logical area. This approach makes it easier to maintain the module.
```
Comment (joelhans) [inline docs/how-to-write-a-module.md:149]:
```suggestion
We recommend using one file per logical area. This approach makes it easier to maintain the module.
```

**Review by joelhans — COMMENTED** [review]
I've talked about this with @OdysLam a few times, but here's my general feedback.

I think there's a lot more that we could do, inside of this document and beyond, to simplify the process of contributors developing their own Go collector. I could go on forever making tiny fixes to this PR as it exists now, but I don't see a ton of value in that for anyone.

I'd rather we publish, share with users, and solicit feedback on how to improve the guide based on their experience. I can give it a pass for sentence-level clarity in the next week or two as a follow-up.

**Comment by OdysLam** [comment]
I agree, let's ship this @ilyam8 @joelhans
